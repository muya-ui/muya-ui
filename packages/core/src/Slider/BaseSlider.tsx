import React, { useMemo } from 'react';

import Tooltip from '../Tooltip';
import Typography from '../Typography';
import memoForwardRef from '../utils/memoForwardRef';
import mergeStyleItem from '../utils/mergeStyleItem';
import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';

import { IBaseSliderProps } from './innerTypes';
import {
  StyledBaseRootTrack,
  StyledCircle,
  StyledMarkLabel,
  StyledMarkPoint,
  StyledMarksRow,
  StyledRange,
  StyledRoot,
} from './styled';
import { ISliderMarkItem } from './types';
import useBaseSlider from './useBaseSlider';

function defaultTipFormatter(str: number) {
  return str.toString();
}

const defaultStyles = {
  startCircle: '',
  endCircle: '',
  markPoint: '',
  markPointInclude: '',
  marksRow: '',
  markLabel: '',
  track: '',
  range: '',
};
const BaseSlider = memoForwardRef<HTMLDivElement, IBaseSliderProps>((props, ref) => {
  const {
    startCircleStyle,
    endCircleStyle,
    rangeStyle,
    start,
    end,

    innerMarksOnly,
    finalMarks,
    markPropsMap,
    markNums,
    movingType,
    trackRef,
    handleStartMoveStartCircle,
    handleStartMoveEndCircle,
    handleTrackClick,

    handleEnterStart,
    handleLeaveStart,
    handleEnterEnd,
    handleLeaveEnd,

    endTooltipOpen,
    startTooltipOpen,

    // 透传
    hideStart,
    vertical,
    tooltipVisible,
    otherProps,
    styles,
    disabled,
    tooltipPlacement,
    markLabelDisabled,
    tipFormatter = defaultTipFormatter,
  } = useBaseSlider(props);
  const innerStyles = useStyles('slider', defaultStyles, styles);
  const theme = useTheme();
  const memoStartCircleStyle = useMemo(() => {
    if (!innerStyles.startCircle.style) {
      return startCircleStyle;
    }
    return {
      ...innerStyles.startCircle.style,
      ...startCircleStyle,
    };
  }, [innerStyles.startCircle.style, startCircleStyle]);
  const startCircleNode = useMemo(
    () =>
      hideStart ? null : (
        <StyledCircle
          className={innerStyles.startCircle.className}
          $moving={movingType === 'start'}
          theme={theme}
          $disabled={disabled}
          $vertical={vertical}
          style={memoStartCircleStyle}
          onMouseDown={handleStartMoveStartCircle}
          onMouseEnter={handleEnterStart}
          onMouseLeave={handleLeaveStart}
        />
      ),
    [
      disabled,
      handleEnterStart,
      handleLeaveStart,
      handleStartMoveStartCircle,
      hideStart,
      innerStyles.startCircle.className,
      memoStartCircleStyle,
      movingType,
      theme,
      vertical,
    ],
  );
  const memoEndCircleStyle = useMemo(() => {
    if (!innerStyles.endCircle.style) {
      return endCircleStyle;
    }
    return {
      ...innerStyles.endCircle.style,
      ...endCircleStyle,
    };
  }, [endCircleStyle, innerStyles.endCircle.style]);
  // 第二个控制点的渲染
  const endCircleNode = useMemo(
    () => (
      <StyledCircle
        className={innerStyles.endCircle.className}
        $moving={movingType === 'end'}
        theme={theme}
        $disabled={disabled}
        $vertical={vertical}
        style={memoEndCircleStyle}
        onMouseDown={handleStartMoveEndCircle}
        onMouseEnter={handleEnterEnd}
        onMouseLeave={handleLeaveEnd}
      />
    ),
    [
      disabled,
      handleEnterEnd,
      handleLeaveEnd,
      handleStartMoveEndCircle,
      innerStyles.endCircle.className,
      memoEndCircleStyle,
      movingType,
      theme,
      vertical,
    ],
  );
  let startNode = startCircleNode;
  let endNode = endCircleNode;
  if (tooltipVisible) {
    startNode = startCircleNode && (
      <Tooltip title={tipFormatter(start)} placement={tooltipPlacement} open={startTooltipOpen}>
        {startCircleNode}
      </Tooltip>
    );
    endNode = (
      <Tooltip title={tipFormatter(end)} placement={tooltipPlacement} open={endTooltipOpen}>
        {endCircleNode}
      </Tooltip>
    );
  }
  let markPointsNode;
  let marksRowNode;

  if (finalMarks) {
    markPointsNode = markNums.map(keyNum => {
      const $include = keyNum > start && keyNum < end;
      const markProps = markPropsMap[keyNum];
      let innerStyle = markProps.style;
      if ($include && innerStyles.markPointInclude.style) {
        innerStyle = {
          ...innerStyles.markPointInclude.style,
          ...innerStyle,
        };
      } else if (!$include && innerStyles.markPoint.style) {
        innerStyle = {
          ...innerStyles.markPoint.style,
          ...innerStyle,
        };
      }
      const pointClassName = $include
        ? innerStyles.markPointInclude.className
        : innerStyles.markPoint.className;

      return (
        <StyledMarkPoint
          className={pointClassName}
          theme={theme}
          $include={$include}
          $disabled={disabled}
          key={keyNum}
          $vertical={vertical}
          style={innerStyle}
          onClick={markProps.onClick}
        />
      );
    });
  }
  const markLabelPropsMap = useMemo(() => {
    const map: Record<
      number,
      { markLabel: React.ReactNode; markProps: React.HTMLAttributes<HTMLDivElement> }
    > = {};
    if (!finalMarks) {
      return map;
    }
    markNums.forEach(keyNum => {
      const label = finalMarks[keyNum];
      const markItem = label as ISliderMarkItem;
      const markProps = markPropsMap[keyNum];
      let markStyle = markProps.style;
      let markLabel = label;
      if (markItem.label) {
        markLabel = markItem.label;
      }
      if (markItem.style) {
        markStyle = {
          ...markItem.style,
          ...markStyle,
        };
      }
      if (innerStyles.markLabel.style) {
        markStyle = {
          ...innerStyles.markLabel.style,
          ...markStyle,
        };
      }
      map[keyNum] = {
        markProps: {
          className: innerStyles.markLabel.className,
          style: markStyle,
          onClick: markProps.onClick,
        },
        markLabel,
      };
    });
    return map;
  }, [
    finalMarks,
    innerStyles.markLabel.className,
    innerStyles.markLabel.style,
    markNums,
    markPropsMap,
  ]);
  if (finalMarks && !markLabelDisabled) {
    marksRowNode = (
      <StyledMarksRow theme={theme} $vertical={vertical}>
        {markNums.map(keyNum => {
          const { markProps, markLabel } = markLabelPropsMap[keyNum];
          let labelNode;
          if (React.isValidElement(markLabel)) {
            labelNode = markLabel;
          } else {
            labelNode = <Typography.Text>{markLabel}</Typography.Text>;
          }

          return (
            <StyledMarkLabel
              theme={theme}
              $disabled={disabled}
              key={keyNum}
              $vertical={vertical}
              {...markProps}
            >
              {labelNode}
            </StyledMarkLabel>
          );
        })}
      </StyledMarksRow>
    );
  }

  const rangeStyleItem = useMemo(() => mergeStyleItem({ style: rangeStyle }, innerStyles.range), [
    innerStyles.range,
    rangeStyle,
  ]);

  return (
    <StyledRoot
      theme={theme}
      $disabled={disabled}
      $vertical={vertical}
      $moving={movingType !== 'none'}
      $disableClick={innerMarksOnly}
      {...otherProps}
      ref={ref}
    >
      <StyledBaseRootTrack
        {...innerStyles.track}
        theme={theme}
        $disabled={disabled}
        $disableClick={innerMarksOnly}
        $moving={movingType !== 'none'}
        $vertical={vertical}
        ref={trackRef}
        onClick={handleTrackClick}
      >
        {startNode}
        {endNode}
        {markPointsNode}
        <StyledRange
          $moving={movingType !== 'none'}
          theme={theme}
          $disabled={disabled}
          $vertical={vertical}
          $disableClick={innerMarksOnly}
          {...rangeStyleItem}
        />
      </StyledBaseRootTrack>
      {marksRowNode}
    </StyledRoot>
  );
});

export default BaseSlider;
