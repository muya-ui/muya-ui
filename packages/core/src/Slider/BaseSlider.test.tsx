import React from 'react';
import renderer from 'react-test-renderer';

import BaseSlider from './BaseSlider';

test('测试普通渲染', () => {
  const tree = renderer
    .create(
      <>
        <BaseSlider />
        <BaseSlider step={10} />
        <BaseSlider tooltipVisible />
        <BaseSlider
          marks={{
            10: 'sss',
          }}
        />
        <BaseSlider
          marks={{
            10: {
              style: {
                color: 'red',
              },
              label: '1100',
            },
          }}
        />
        <BaseSlider
          value={[0, 20]}
          styles={{
            markLabel: {
              color: 'blue',
            },
            startCircle: {
              color: 'blue',
            },
            endCircle: {
              color: 'blue',
            },
            markPointInclude: {
              color: 'blue',
            },
            markPoint: {
              color: 'blue',
            },
          }}
          marks={{
            10: {
              style: {
                color: 'red',
              },
              label: '1100',
            },
            30: {
              style: {
                color: 'red',
              },
              label: '1100',
            },
          }}
        />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 disabled', () => {
  const tree = renderer.create(<BaseSlider disabled />).toJSON();
  expect(tree).toMatchSnapshot();
});
