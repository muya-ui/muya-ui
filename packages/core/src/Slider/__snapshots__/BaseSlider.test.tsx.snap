// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 disabled 1`] = `
.c5:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c0 {
  position: relative;
  box-sizing: border-box;
  cursor: not-allowed;
  padding: 4px 0;
}

.c4 {
  position: absolute;
  cursor: not-allowed;
  border-radius: 999rem;
  background-color: rgba(221,221,221,1);
  z-index: 1;
  height: 100%;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: not-allowed;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(221,221,221,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: not-allowed;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
  className="c0"
>
  <div
    className="c1 c2 muya-slider-track"
    onClick={[Function]}
  >
    <div
      className="c3 muya-slider-start-circle"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "left": "0%",
        }
      }
    />
    <div
      className="c3 muya-slider-end-circle"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "left": "0%",
        }
      }
    />
    <div
      className="c4 muya-slider-range"
      style={
        Object {
          "left": "0%",
        }
      }
    />
  </div>
</div>
`;

exports[`测试普通渲染 1`] = `
Array [
  .c0 {
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  padding: 4px 0;
}

.c0:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c4 {
  position: absolute;
  cursor: pointer;
  border-radius: 999rem;
  background-color: rgba(26,122,248,1);
  z-index: 1;
  height: 100%;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: pointer;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(26,122,248,1);
}

.c3:hover {
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 16px;
  height: 16px;
  box-shadow: none;
}

.c3:hover::before {
  border-color: rgba(26,122,248,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
    className="c0"
  >
    <div
      className="c1 c2 muya-slider-track"
      onClick={[Function]}
    >
      <div
        className="c3 muya-slider-start-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c3 muya-slider-end-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c4 muya-slider-range"
        style={
          Object {
            "left": "0%",
          }
        }
      />
    </div>
  </div>,
  .c0 {
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  padding: 4px 0;
}

.c0:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c4 {
  position: absolute;
  cursor: pointer;
  border-radius: 999rem;
  background-color: rgba(26,122,248,1);
  z-index: 1;
  height: 100%;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: pointer;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(26,122,248,1);
}

.c3:hover {
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 16px;
  height: 16px;
  box-shadow: none;
}

.c3:hover::before {
  border-color: rgba(26,122,248,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
    className="c0"
  >
    <div
      className="c1 c2 muya-slider-track"
      onClick={[Function]}
    >
      <div
        className="c3 muya-slider-start-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c3 muya-slider-end-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c4 muya-slider-range"
        style={
          Object {
            "left": "0%",
          }
        }
      />
    </div>
  </div>,
  .c0 {
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  padding: 4px 0;
}

.c0:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c4 {
  position: absolute;
  cursor: pointer;
  border-radius: 999rem;
  background-color: rgba(26,122,248,1);
  z-index: 1;
  height: 100%;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: pointer;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(26,122,248,1);
}

.c3:hover {
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 16px;
  height: 16px;
  box-shadow: none;
}

.c3:hover::before {
  border-color: rgba(26,122,248,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
    className="c0"
  >
    <div
      className="c1 c2 muya-slider-track"
      onClick={[Function]}
    >
      <div
        aria-describedby={null}
        className="c3 muya-slider-start-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        aria-describedby={null}
        className="c3 muya-slider-end-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c4 muya-slider-range"
        style={
          Object {
            "left": "0%",
          }
        }
      />
    </div>
  </div>,
  .c8 {
  color: rgba(102,102,102,1);
  font-size: 12px;
  line-height: 16px;
  box-sizing: border-box;
  margin: 0;
}

.c0 {
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  padding: 4px 0;
}

.c0:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c5 {
  position: absolute;
  cursor: pointer;
  border-radius: 999rem;
  background-color: rgba(26,122,248,1);
  z-index: 1;
  height: 100%;
}

.c4 {
  position: absolute;
  box-sizing: border-box;
  cursor: pointer;
  z-index: 2;
  height: 12px;
  width: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c4::before {
  content: '';
  display: block;
  box-sizing: border-box;
  height: 6px;
  width: 6px;
  border: 1px solid rgba(221,221,221,1);
  border-radius: 999rem;
  background-color: rgba(255,255,255,1);
}

.c6 {
  position: relative;
  width: 100%;
  margin-top: 8px;
  height: 20px;
}

.c7 {
  position: absolute;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
  cursor: pointer;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
  top: 0;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: pointer;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(26,122,248,1);
}

.c3:hover {
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 16px;
  height: 16px;
  box-shadow: none;
}

.c3:hover::before {
  border-color: rgba(26,122,248,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
    className="c0"
  >
    <div
      className="c1 c2 muya-slider-track"
      onClick={[Function]}
    >
      <div
        className="c3 muya-slider-start-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c3 muya-slider-end-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c4 muya-slider-mark-point"
        onClick={[Function]}
        style={
          Object {
            "left": "10%",
          }
        }
      />
      <div
        className="c5 muya-slider-range"
        style={
          Object {
            "left": "0%",
          }
        }
      />
    </div>
    <div
      className="c6"
    >
      <div
        className="c7 muya-slider-mark-label"
        onClick={[Function]}
        style={
          Object {
            "left": "10%",
          }
        }
      >
        <span
          className="c8"
        >
          sss
        </span>
      </div>
    </div>
  </div>,
  .c8 {
  color: rgba(102,102,102,1);
  font-size: 12px;
  line-height: 16px;
  box-sizing: border-box;
  margin: 0;
}

.c0 {
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  padding: 4px 0;
}

.c0:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c5 {
  position: absolute;
  cursor: pointer;
  border-radius: 999rem;
  background-color: rgba(26,122,248,1);
  z-index: 1;
  height: 100%;
}

.c4 {
  position: absolute;
  box-sizing: border-box;
  cursor: pointer;
  z-index: 2;
  height: 12px;
  width: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c4::before {
  content: '';
  display: block;
  box-sizing: border-box;
  height: 6px;
  width: 6px;
  border: 1px solid rgba(221,221,221,1);
  border-radius: 999rem;
  background-color: rgba(255,255,255,1);
}

.c6 {
  position: relative;
  width: 100%;
  margin-top: 8px;
  height: 20px;
}

.c7 {
  position: absolute;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
  cursor: pointer;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
  top: 0;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: pointer;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(26,122,248,1);
}

.c3:hover {
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 16px;
  height: 16px;
  box-shadow: none;
}

.c3:hover::before {
  border-color: rgba(26,122,248,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
    className="c0"
  >
    <div
      className="c1 c2 muya-slider-track"
      onClick={[Function]}
    >
      <div
        className="c3 muya-slider-start-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c3 muya-slider-end-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <div
        className="c4 muya-slider-mark-point"
        onClick={[Function]}
        style={
          Object {
            "left": "10%",
          }
        }
      />
      <div
        className="c5 muya-slider-range"
        style={
          Object {
            "left": "0%",
          }
        }
      />
    </div>
    <div
      className="c6"
    >
      <div
        className="c7 muya-slider-mark-label"
        onClick={[Function]}
        style={
          Object {
            "color": "red",
            "left": "10%",
          }
        }
      >
        <span
          className="c8"
        >
          1100
        </span>
      </div>
    </div>
  </div>,
  .c9 {
  color: rgba(102,102,102,1);
  font-size: 12px;
  line-height: 16px;
  box-sizing: border-box;
  margin: 0;
}

.c0 {
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  padding: 4px 0;
}

.c0:hover .c1 {
  background-color: rgba(242,242,242,1);
}

.c6 {
  position: absolute;
  cursor: pointer;
  border-radius: 999rem;
  background-color: rgba(26,122,248,1);
  z-index: 1;
  height: 100%;
}

.c5 {
  position: absolute;
  box-sizing: border-box;
  cursor: pointer;
  z-index: 2;
  height: 12px;
  width: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c5::before {
  content: '';
  display: block;
  box-sizing: border-box;
  height: 6px;
  width: 6px;
  border: 1px solid rgba(221,221,221,1);
  border-radius: 999rem;
  background-color: rgba(255,255,255,1);
}

.c4 {
  position: absolute;
  box-sizing: border-box;
  cursor: pointer;
  z-index: 2;
  height: 12px;
  width: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c4::before {
  content: '';
  display: block;
  box-sizing: border-box;
  height: 6px;
  width: 6px;
  border: 1px solid rgba(26,122,248,1);
  border-radius: 999rem;
  background-color: rgba(255,255,255,1);
}

.c7 {
  position: relative;
  width: 100%;
  margin-top: 8px;
  height: 20px;
}

.c8 {
  position: absolute;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
  cursor: pointer;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
  top: 0;
}

.c3 {
  position: absolute;
  z-index: 3;
  box-sizing: content-box;
  cursor: pointer;
  background-color: rgba(255,255,255,1);
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
}

.c3::before {
  content: '';
  height: 100%;
  width: 100%;
  display: block;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: rgba(255,255,255,1);
  border: 1px solid rgba(26,122,248,1);
}

.c3:hover {
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 16px;
  height: 16px;
  box-shadow: none;
}

.c3:hover::before {
  border-color: rgba(26,122,248,1);
}

.c2 {
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  background-color: rgba(242,242,242,1);
  border-radius: 999rem;
  width: 100%;
  height: 6px;
}

<div
    className="c0"
  >
    <div
      className="c1 c2 muya-slider-track"
      onClick={[Function]}
    >
      <div
        className="c3 muya-slider-start-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "blue",
            "left": "0%",
          }
        }
      />
      <div
        className="c3 muya-slider-end-circle"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "blue",
            "left": "20%",
          }
        }
      />
      <div
        className="c4 muya-slider-mark-point-include"
        onClick={[Function]}
        style={
          Object {
            "color": "blue",
            "left": "10%",
          }
        }
      />
      <div
        className="c5 muya-slider-mark-point"
        onClick={[Function]}
        style={
          Object {
            "color": "blue",
            "left": "30%",
          }
        }
      />
      <div
        className="c6 muya-slider-range"
        style={
          Object {
            "left": "0%",
            "width": "20%",
          }
        }
      />
    </div>
    <div
      className="c7"
    >
      <div
        className="c8 muya-slider-mark-label"
        onClick={[Function]}
        style={
          Object {
            "color": "red",
            "left": "10%",
          }
        }
      >
        <span
          className="c9"
        >
          1100
        </span>
      </div>
      <div
        className="c8 muya-slider-mark-label"
        onClick={[Function]}
        style={
          Object {
            "color": "red",
            "left": "30%",
          }
        }
      >
        <span
          className="c9"
        >
          1100
        </span>
      </div>
    </div>
  </div>,
]
`;
